### PR TITLE
🐛 Ensure controlplane coredns update deploys the ClusterRole if CoreDNS was already updated

### LINKED ISSUE #6676: CAPI periodic-cluster-api-e2e-workload-upgrade-1-21-1-22-main job fails frequently
According to testgrid the 1.21=>1.22 upgrade job fails frequently: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-main-1-21-1-22

The error looks like this
> E0620 11:02:59.604086       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.21.1/tools/cache/reflector.go:167: Failed to watch *v1.EndpointSlice: failed to list *v1.EndpointSlice: endpointslices.discovery.k8s.io is forbidden: User "system:serviceaccount:kube-system:coredns" cannot list resource "endpointslices" in API group "discovery.k8s.io" at the cluster scope

My theory is that somehow we don't patch the CoreDNS role in those test runs. Usually we patch it for Kubernetes >=1.22 and CoreDNS >= 1.8.1.

https://github.com/kubernetes-sigs/cluster-api/blob/5026786ee809c5def466049f5befd2a786fbcefa/controlplane/kubeadm/internal/workload_cluster_coredns.go#L261-L291

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by chrischdi**
… was already updated

**What this PR does / why we need it**:

Adjusts the core DNS update logic to always ensure the correct ClusterRole.

Following scenario:
* A cluster gets provisioned with kubernetes v1.21 and CoreDNS >= v1.8.1
* The CoreDNS ClusterRole will not contain EndpointSlices (because kubeadm won't add them and neither the KCP controller for kubernetes <= v1.21, which is expected)
* The cluster then gets updated to v1.22 referencing the same CoreDNS version
* The old logic won't update the CoreDNS ClusterRole to contain EndpointSlices because the version does not get updated.
* CoreDNS will not be ready anymore because it tries to access EndpointSlices but is not allowed to by RBAC.

This scenario could happen when using ClusterClass and e.g.:
* A `BeforeControlPlane` hook which currently blocks (`RetryAfterSeconds`)
* A MachineDeployment is in scaling

TLDR for ClusterClass: if [computeControlPlaneVersion] returns `currentVersion` (`v1.21.X`) instead of desiredVersion (`v1.22.X`) and thus CoreDNS gets upgraded without ClusterRole update.

[computeControlPlaneVersion]: https://github.com/kubernetes-sigs/cluster-api/blob/c74b04c8252abc374ac6330099a533f9eb7771da/internal/controllers/topology/cluster/desired_state.go#L260

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6676

### CODE REVIEW

File: controlplane/kubeadm/internal/workload_cluster_coredns.go [hunk 1]
```diff
@@ -125,13 +126,18 @@ func (w *Workload) UpdateCoreDNS(ctx context.Context, kcp *controlplanev1.Kubead
 		return err
 	}
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster_coredns.go:127]: Is there a way to do this without having to "get" the ClusterRole every time we execute UpdateCoreDNS (which is afaik on every KCP reconcile)?

Or, maybe it doesn't matter because we are caching all ClusterRoles currently anyway? (IIRC we don't set ClusterRole as uncached so the Get call should give us an informer & cache currently anyway?)

File: controlplane/kubeadm/internal/workload_cluster_coredns.go [hunk 2]
```diff
@@ -125,13 +126,18 @@ func (w *Workload) UpdateCoreDNS(ctx context.Context, kcp *controlplanev1.Kubead
...
+	}
+
+	// Return early if the from/to image is the same and ClusterRole is up to date.
+	if info.FromImage == info.ToImage && !requiresClusterRoleUpdate {
 		return nil
 	}
 
 	// Validate the image tag.
-	if err := validateCoreDNSImageTag(info.FromImageTag, info.ToImageTag); err != nil {
+	if err := validateCoreDNSImageTag(info.FromImageTag, info.ToImageTag, requiresClusterRoleUpdate); err != nil {
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster_coredns.go:140]: Can we maybe only call this func if .FromImage != ToImage? This should avoid making the logic in the func even messier :)

Comment (sbueringer) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go:140]: Overall I wonder if we can structure this a bit cleaner.
I.e. call certain stuff if FromImage != ToImage and call certain stuff if requiresClusterRoleUpdate (or maybe some of them if one or the other is true)

File: controlplane/kubeadm/internal/workload_cluster_coredns.go (hunk 1 above)
Comment (chrischdi) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go:127]: I don't see any other way. And yes, the ClusterRole is cached so I think this is ok.

Comment (sbueringer) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go:127]: Hm we could skip it if Kubernetes >= v1.23 and CoreDNS FromImageTag >= 1.8.1 I guess?

This would mean that the ClusterRole was patched before. Not sure if it makes the logic too complicated. But it would mean that for most current clusters we would never execute the code and we would never create an informer+cache for the ClusterRole.

(which essentially means going forward this will be never done)

Comment (chrischdi) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go:127]: 👍  to skipping for >= v1.23 .

CoreDNS FromImageTag >= 1.8.1,no because:

* k8s v1.21 + coredns >= 1.8.1 -> no endpointslices
* k8s v1.22 + coredns >= 1.8.1 (keep same coredns version) -> endpointslices

Comment (chrischdi) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go:126]: Logic is now wway simpler :-)

File: controlplane/kubeadm/internal/workload_cluster_coredns.go [hunk 3]
```diff
@@ -273,6 +278,12 @@ func (w *Workload) updateCoreDNSClusterRole(ctx context.Context, kubernetesVersi
 		return nil
 	}
 
+	// Do nothing for Kubernetes > 1.22 and coreDNS Version > 1.8.1.
+	if kubernetesVersion.GTE(semver.Version{Major: 1, Minor: 23, Patch: 0}) &&
+		targetCoreDNSVersion.GTE(semver.Version{Major: 1, Minor: 8, Patch: 1}) {
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster_coredns.go:283]: I think for this if we have to use the sourceCoreDNSVersion so we know that we previously upgraded to a CoreDNS version >= v1.8.1

**Comment by chrischdi** [comment]
Last test was green, we want to fix a flaky test so let's try again:

/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main

**Comment by chrischdi** [comment]
Looks good so far, last was green, testing again:

/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go [hunk 4]
```diff
@@ -160,25 +204,33 @@ func TestUpdateCoreDNS(t *testing.T) {
 					KubeadmConfigSpec: bootstrapv1.KubeadmConfigSpec{},
 				},
 			},
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster_coredns_test.go:206]: Q: could it be that we don't have to change so many of the test cases given that we changed the code again?

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go [hunk 5]
```diff
@@ -382,12 +450,40 @@ func TestUpdateCoreDNS(t *testing.T) {
...
+			migrator: &fakeMigrator{
+				migratedCorefile: "updated-core-file",
+			},
+			semver:        semver1221,
+			objs:          []client.Object{deplWithImage("k8s.gcr.io/coredns/coredns:v1.8.1"), updatedCM, kubeadmCM181, oldCR},
+			expectErr:     false,
+			expectUpdates: true,
+			expectImage:   "k8s.gcr.io/coredns/coredns:v1.8.1", // NOTE: ImageName has coredns/coredns
+			expectRules:   coreDNS181PolicyRules,
 		},
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster_coredns_test.go:487]: Maybe add a test case for >=v1.23 && CoreDNS sourceVersion >= 1.8.1 where we shouldn't update the rules

**Comment by sbueringer** [comment]
I think it should be fine, expect the one case I commented above

File: controlplane/kubeadm/internal/workload_cluster_coredns.go [hunk 6]
```diff
@@ -125,28 +125,33 @@ func (w *Workload) UpdateCoreDNS(ctx context.Context, kcp *controlplanev1.Kubead
...
 	}
 
-	// Perform the upgrade.
-	if err := w.updateCoreDNSImageInfoInKubeadmConfigMap(ctx, &clusterConfig.DNS, version); err != nil {
-		return err
-	}
-	if err := w.updateCoreDNSCorefile(ctx, info); err != nil {
-		return err
-	}
+	// Update the cluster role independent of image change. Kubernetes may get updated
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/workload_cluster_coredns.go:144]: What about moving this up L128, before all the changes that depends on the image changing

Comment (chrischdi) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go:144]: I did not want to change the order of actions. But I'd be also ok with that :-) I think as long as it is still before a potential deployment change this should be fine.

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go (hunk 4 above)
Comment (chrischdi) [reply controlplane/kubeadm/internal/workload_cluster_coredns_test.go:207]: ack, reverted the unnecessary parts.

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go (hunk 5 above)
Comment (chrischdi) [reply controlplane/kubeadm/internal/workload_cluster_coredns_test.go:488]: done

File: controlplane/kubeadm/internal/workload_cluster_coredns.go (hunk 6 above)
Comment (sbueringer) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go:128]: Yup +1 before is fine for me

Comment (chrischdi) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go:128]: done, and also resolved the test issues.

File: controlplane/kubeadm/internal/workload_cluster_coredns.go [hunk 7]
```diff
@@ -273,6 +277,12 @@ func (w *Workload) updateCoreDNSClusterRole(ctx context.Context, kubernetesVersi
 		return nil
 	}
 
+	// Do nothing for Kubernetes > 1.22 and coreDNS Version > 1.8.1.
+	if kubernetesVersion.GTE(semver.Version{Major: 1, Minor: 23, Patch: 0}) &&
+		targetCoreDNSVersion.GTE(semver.Version{Major: 1, Minor: 8, Patch: 1}) {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go]: Adjusted 👍

Test which matches does already exists.

**Comment by sbueringer** [comment]
Prod code looks good, tests need some fixes

**Comment by fabriziopandini** [comment]
Lgtm pending CI green signal

**Comment by chrischdi** [comment]
This should fix ci signal. I forgot to add the source coreDNSVersion which is now also used in `updateCoreDNSClusterRole(...)` to the test cases.

**Comment by sbueringer** [comment]
1.24=>latest failure is expected as it is currently broken

/test pull-cluster-api-e2e-workload-upgrade-1-21-1-22-main
/test pull-cluster-api-e2e-full-main
